apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1065 @@
+(the trust-on-first-use paradigm), permit users to manually select variants
+or disable variant use entirely. For example, a user who does not have an NVIDIA
+GPU does not really need to query the respective plugin.
+
+## The present, and the future
+
+In this post, I've attempted to comment on the road the variant wheel work
+has taken from the project's inception to the current pre-proto-PEP: [PEP ### -
+Wheel Variants](http://variant-proposal.wheelnext.dev/).
+Necessarily, it is neither final nor complete in all the detail, but I think
+it does some justice to the complexity of the problem space and the effort
+put into developing a good solution.
+
+Overall, we have been following [the philosophy of WheelNext
+project](https://wheelnext.dev/philosophy_and_design_space/), and focusing
+on providing a working solutions to real problems. I believe that this approach
+worked well, and we arrived at a proposal that does not introduce more
+complexity than absolutely necessary to solve the problems we were facing,
+and attempts to use generic solutions. More than once, it turned out that
+a single past decision either solved future problems that we did not originally
+anticipate, or made a solution possible with minimal changes to the design.
+
+We started with a simple conjunctive property framework, managed to easily
+handle flag-like properties with it and then extended it to handle multiple
+values disjunctively — solving the GPU target problem with minimal changes
+to the specification. We introduced `variants.json` to solve an immediate
+scalability problem, yet it enabled us to easily introduce custom variant labels
+later on. We created a relatively simple and clean API, and were able
+to extend it to support dynamic plugins with minimal changes. And the sort
+algorithm originally devised for the original one-value-per-property variants,
+requires no changes when we introduced multi-value properties and the null
+variant.
+
+Of course, the work is far from complete. We have solved a number of problems,
+we have arrived at a working prototype — but as more people start testing
+the prototype against their own packages, and as more people join
+the discussion, new use cases and new problems will emerge. This does
+not necessarily mean that we will have to go back to the proverbial drawing
+board. It is entirely possible that some of these properties can be solved
+within the existing framework, much like we managed to implement a good enough
+CUDA plugin without having to introduce a dynamic plugin, or that they will
+turn out not to be worth the added complexity. It is also possible that
+some of the existing features will be removed.
+
+There are open issues still. There's the question of how lockfiles are supposed
+to work with variants. There's the possible matter of dependencies on specific
+variants. But most importantly, wheel variants opened a Pandora's box
+of potential security issues that we need to find good mitigations for.
```

Comment:
I'm not too enthusiastic about this sentence either:)

Reply:
Yeah, I might have tried to be too metaphoric about it :-).